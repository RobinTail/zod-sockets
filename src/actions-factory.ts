import { z } from "zod";
import { Action, Handler } from "./action";
import { EmissionMap } from "./emission";
import { SocketsConfig } from "./sockets";

export interface SimpleActionDef<
  IN extends z.AnyZodTuple,
  E extends EmissionMap,
> {
  input: IN;
  handler: Handler<z.output<IN>, void, E>;
}

export interface AckActionDef<
  IN extends z.AnyZodTuple,
  OUT extends z.AnyZodTuple,
  E extends EmissionMap,
> {
  input: IN;
  output: OUT;
  handler: Handler<z.output<IN>, z.input<OUT>, E>;
}

export class ActionsFactory<E extends EmissionMap> {
  constructor(protected config: SocketsConfig<E>) {}

  public build<IN extends z.AnyZodTuple, OUT extends z.AnyZodTuple>(
    def: SimpleActionDef<IN, E> | AckActionDef<IN, OUT, E>,
  ): Action<IN, OUT> {
    return new Action(def);
  }
}
