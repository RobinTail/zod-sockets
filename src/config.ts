import { EmissionMap } from "./emission";
import { AbstractLogger } from "./logger";
import {
  FallbackNamespaces,
  Namespace,
  Namespaces,
  RootNS,
  fallbackNamespaces,
  rootNS,
} from "./namespace";

interface ConstructorOptions<NS extends Namespaces> {
  /**
   * @desc The instance of a logger
   * @default console
   * */
  logger?: AbstractLogger;
  /**
   * @desc The acknowledgment awaiting timeout
   * @default 2000
   * */
  timeout?: number;
  /**
   * @desc You can disable the startup logo.
   * @default true
   */
  startupLogo?: boolean;
  /**
   * @desc Define namespaces inline or consider using addNamespace() method
   * @default {}
   * @see Namespace
   * */
  namespaces?: NS;
}

export class Config<T extends Namespaces> {
  public readonly logger: AbstractLogger;
  public readonly timeout: number;
  public readonly startupLogo: boolean;
  public readonly namespaces: T;

  public constructor({
    logger = console,
    timeout = 2000,
    startupLogo = true,
    namespaces = fallbackNamespaces as unknown as T,
  }: ConstructorOptions<T>) {
    this.logger = logger;
    this.timeout = timeout;
    this.startupLogo = startupLogo;
    this.namespaces = namespaces;
  }

  public addNamespace<E extends EmissionMap = {}, K extends string = RootNS>({
    path = rootNS as K,
    emission = {} as E,
    hooks = {},
  }: Partial<Namespace<E>> & { path?: K }): Config<
    Omit<T, K> & Record<K, Namespace<E>>
  > {
    const { logger, timeout, startupLogo, namespaces } = this;
    return new Config({
      logger,
      timeout,
      startupLogo,
      namespaces: { ...namespaces, [path]: { emission, hooks } },
    });
  }
}

export const createConfig = <T extends Namespaces = FallbackNamespaces>(
  def: ConstructorOptions<T> = {},
) => new Config(def);
