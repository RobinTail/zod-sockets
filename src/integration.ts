import ts from "typescript";
import { z } from "zod";
import { AbstractAction } from "./action";
import { Config } from "./config";
import { EmissionMap } from "./emission";
import {
  defaultSerializer,
  exportModifier,
  f,
  makeCleanId,
  makeEventFnSchema,
} from "./integration-helpers";
import { Namespaces } from "./namespaces";
import { zodToTs } from "./zts";
import { addJsDocComment, createTypeAlias, printNode } from "./zts-helpers";

interface IntegrationProps {
  config: Config<Namespaces<EmissionMap>>;
  actions: AbstractAction[];
  /**
   * @desc Used for comparing schemas wrapped into z.lazy() to limit the recursion
   * @default JSON.stringify() + SHA1 hash as a hex digest
   * */
  serializer?: (schema: z.ZodTypeAny) => string;
  /**
   * @desc configures the style of object's optional properties
   * @default { withQuestionMark: true, withUndefined: true }
   */
  optionalPropStyle?: {
    /**
     * @desc add question mark to the optional property definition
     * @example { someProp?: boolean }
     * */
    withQuestionMark?: boolean;
    /**
     * @desc add undefined to the property union type
     * @example { someProp: boolean | undefined }
     */
    withUndefined?: boolean;
  };
}

export class Integration {
  protected program: ts.Node[] = [];
  protected aliases: Record<string, ts.TypeAliasDeclaration> = {};
  protected registry: Record<
    string, // namespace
    Record<"emission" | "actions", { event: string; node: ts.TypeNode }[]>
  > = {};

  protected getAlias(name: string): ts.TypeReferenceNode | undefined {
    return name in this.aliases ? f.createTypeReferenceNode(name) : undefined;
  }

  protected makeAlias(name: string, type: ts.TypeNode): ts.TypeReferenceNode {
    this.aliases[name] = createTypeAlias(type, name);
    return this.getAlias(name)!;
  }

  constructor({
    config: { emission: namespaces },
    actions,
    serializer = defaultSerializer,
    optionalPropStyle = { withQuestionMark: true, withUndefined: true },
  }: IntegrationProps) {
    this.program.push(
      f.createImportDeclaration(
        undefined,
        f.createImportClause(
          true,
          undefined,
          f.createNamedImports([
            f.createImportSpecifier(
              false,
              f.createIdentifier("Socket"),
              f.createIdentifier("SocketBase"),
            ),
          ]),
        ),
        f.createStringLiteral("socket.io-client"),
        undefined,
      ),
    );

    for (const [ns, emission] of Object.entries(namespaces)) {
      this.registry[ns] = { emission: [], actions: [] };
      for (const [event, { schema, ack }] of Object.entries(emission)) {
        const node = zodToTs({
          schema: makeEventFnSchema(schema, ack, 3),
          isResponse: true,
          getAlias: this.getAlias.bind(this),
          makeAlias: this.makeAlias.bind(this),
          serializer,
          optionalPropStyle,
        });
        this.registry[ns].emission.push({ event, node });
      }
      for (const action of actions) {
        if (action.getNamespace() !== ns) {
          continue;
        }
        const event = action.getEvent();
        const input = action.getSchema("input");
        const output = action.getSchema("output");
        const node = zodToTs({
          schema: makeEventFnSchema(input, output, 3),
          isResponse: false,
          getAlias: this.getAlias.bind(this),
          makeAlias: this.makeAlias.bind(this),
          serializer,
          optionalPropStyle,
        });
        this.registry[ns].actions.push({ event, node });
      }
    }

    for (const ns in this.registry) {
      const publicName = makeCleanId(ns) || makeCleanId("root");
      const interfaces = Object.entries(this.registry[ns]).map(
        ([direction, events]) =>
          f.createInterfaceDeclaration(
            exportModifier,
            makeCleanId(direction),
            undefined,
            undefined,
            events.map(({ event, node }) =>
              f.createPropertySignature(undefined, event, undefined, node),
            ),
          ),
      );
      const socketNode = f.createTypeAliasDeclaration(
        exportModifier,
        f.createIdentifier("Socket"),
        undefined,
        f.createTypeReferenceNode(f.createIdentifier("SocketBase"), [
          f.createTypeReferenceNode(f.createIdentifier("Emission")),
          f.createTypeReferenceNode(f.createIdentifier("Actions")),
        ]),
      );
      addJsDocComment(
        socketNode,
        `@example const socket: ${publicName}.Socket = io("${ns}")`,
      );
      this.program.push(
        f.createModuleDeclaration(
          exportModifier,
          f.createIdentifier(publicName),
          f.createModuleBlock([...interfaces, socketNode]),
          ts.NodeFlags.Namespace,
        ),
      );
    }
  }

  public print(printerOptions?: ts.PrinterOptions) {
    return this.program
      .map((node, index) =>
        printNode(
          node,
          index < this.program.length
            ? printerOptions
            : { ...printerOptions, omitTrailingSemicolon: true },
        ),
      )
      .join("\n\n");
  }
}
