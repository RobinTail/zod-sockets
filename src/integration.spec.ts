import { z } from "zod/v4";
import { actions } from "../example/actions";
import { config } from "../example/config";
import { ActionsFactory } from "./actions-factory";
import { createSimpleConfig } from "./config";
import { Integration } from "./integration";

describe("Integration", () => {
  describe("print()", () => {
    test("should print the example client side types", () => {
      const instance = new Integration({ config, actions });
      expect(instance.print()).toMatchSnapshot();
    });

    test("should handle circular references", () => {
      const feature = z.object({
        title: z.string(),
        get features() {
          return feature.array();
        },
      });
      const input = z.tuple([feature.array()]);
      const instance = new Integration({
        config: createSimpleConfig(),
        actions: [
          new ActionsFactory(config).build({
            event: "test",
            input,
            handler: vi.fn<any>(),
          }),
        ],
      });
      expect(instance.print()).toMatchSnapshot();
    });
  });
});
