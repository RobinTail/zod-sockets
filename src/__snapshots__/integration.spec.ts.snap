// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Integration > print() > should handle circular references 1`] = `
"import type { Socket as SocketBase } from "socket.io-client";

export namespace Root {
    type Type2048581c137c5b2130eb860e3ae37da196dfc25b = {
        title: string;
        features: Type2048581c137c5b2130eb860e3ae37da196dfc25b;
    }[];
    export interface Emission {
        time: (currentIsoTime: string) => void;
        chat: (message: string, extraInfo: {
            /** the ID of author */
            from: string;
        }) => void;
        rooms: (roomIds: string[]) => void;
    }
    export interface Actions {
        test: (p1: {
            title: string;
            features: Type2048581c137c5b2130eb860e3ae37da196dfc25b;
        }) => void;
    }
    /** @example const socket: Root.Socket = io("/") */
    export type Socket = SocketBase<Emission, Actions>;
}"
`;

exports[`Integration > print() > should print the example client side types 1`] = `
"import type { Socket as SocketBase } from "socket.io-client";

export namespace Root {
    export interface Emission {
        time: (currentIsoTime: string) => void;
        chat: (message: string, extraInfo: {
            /** the ID of author */
            from: string;
        }) => void;
        rooms: (roomIds: string[]) => void;
    }
    export interface Actions {
        chat: (message: string) => void;
        ping: ((cb1: (literally: "pong", ...echo: unknown[]) => void) => void) | ((anything1: unknown, cb2: (literally: "pong", ...echo: unknown[]) => void) => void) | ((anything1: unknown, anything2: unknown, cb3: (literally: "pong", ...echo: unknown[]) => void) => void) | ((anything1: unknown, anything2: unknown, anything3: unknown, cb4: (literally: "pong", ...echo: unknown[]) => void) => void);
        subscribe: (...doesNotMatter: unknown[]) => void;
    }
    /** @example const socket: Root.Socket = io("/") */
    export type Socket = SocketBase<Emission, Actions>;
}"
`;
